Baseline risk meta regression

ndunnewind opened the pull request:
Hi @dmphillippo. As discussed some time ago, this is an initial implementation of baseline risk meta-regression. It's not ready to be merged yet, but I first wanted to ask if you agree with this implementation.

dmphillippo commented:
Thanks @ndunnewind, this is looking really nice so far.

A few thoughts and ideas below. All of this is very much up for discussion! :)

1. You have gone down the route of having a `baseline_risk = TRUE/FALSE` flag, which works well and is simple. However, this does currently restrict the model to having a single regression term for all non-reference treatments. This may be the most common model, but it would be nice to allow more flexible specification, e.g. independent interactions by treatment, or by class.

    a. I had thought about possibly implementing this through the `regression` formula interface, with another "special" for baseline risk, e.g. `.mu`, so you would specify the baseline risk regression as `regression = ~.mu:.trt`. The usual `class_interactions` options would apply to this. *But* this may be far too complicated... I do think that having the `baseline_risk` flag is probably more straightforward.

    b. Instead, we could have another argument to specify the type of interactions with baseline risk, similar to the `class_interactions` argument. E.g. `baseline_interactions = c("independent", "common")`. Perhaps we should refactor/alias `class_interactions` to `trt_interactions`, to better align with this new argument. Or even change `baseline_risk` from a boolean to a string to specify the type of baseline risk model, e.g. `baseline_risk = c("none", "treatment", "class")`.
	
2. The new prior spec `prior_br` is nice; I would probably have shared this with `prior_reg` but this is a little more flexible.

3. Currently only works for agd_arm only models. Should work without changes(?) for models with agd_contrast. Can we get this working for models with ipd too? Is that even a model that would ever be fitted? Perhaps...

4. `calculate_baseline_risk()` is great, this should maybe be an internal function rather than exported?
 
5. Really great to put the `certolizumab` data in. Docs will need fleshing out a little, but we can use this for the vignette.
 
6. Tests in the vignette are excellent, thanks. These will need hardening against stochasticity - they fail at random at the moment due to sampling error. The solution is to run a larger number of iterations when testing.
 
7. Plot functionality I imagine would be along the lines of the [vaccine example](https://dmphillippo.github.io/multinma/articles/example_bcg_vaccine.html), i.e. demonstrate the ggplot code to create the plots, rather than add a new function? Unless this new function could also be used to plot regression terms in general?

8. The `relative_effects()` and `predict()` functions will need updating.

Thank you!

ndunnewind commented:
Thanks for the extensive feedback @dmphillippo!

I removed the `baseline_risk` and `prior_br` arguments from `nma()`. Baseline risk meta-regression can now be specified using `regression = ~.mu:.trt`, like you suggested. The current implementation probably slows down all "normal" models a bit, because I moved some Stan code from `transformed data` to `transformed parameters` to make this implementation work. I'll look into this later to make sure performance of other models is not affected.

For `relative_effects()`, I think the sampled study-specific intercepts should be used by default to calculate study-specific relative effect estimates. A `data.frame` with specific values of `.mu` should be passed as `newdata` to calculate non-study-specific relative effects (see for example the vignette). Do you agree?

The sampled study-specific intercepts should also be used by default for `predict()` to make study-specific absolute predictions. When passing a `baseline` distribution, the sampled values from this distribution should be used as `.mu`, and `.mu` should not be in `newdata`. What do you think?

dmphillippo commented:
Thanks @ndunnewind, this is looking great 🙂 

> The current implementation probably slows down all "normal" models a bit, because I moved some Stan code from transformed data to transformed parameters to make this implementation work.

I see you've moved the design matrices to `transformed parameters`; I am slightly concerned about this slowing down models. Probably there is minimal difference for standard NMAs, but IPD NMA and ML-NMR with all the numerical integration may take a heavier hit. Is there a way to do this keeping the design matrices in `transformed data`? Perhaps you can split off the baseline risk columns of X in `transformed data` and then deal with just these bits in `transformed parameters` and/or `model` blocks? You were previously adding on the baseline risk terms to the linear predictor in the `model` block, I think, which could work?

> For relative_effects(), I think the sampled study-specific intercepts should be used by default to calculate study-specific relative effect estimates. A data.frame with specific values of .mu should be passed as newdata to calculate non-study-specific relative effects (see for example the vignette). Do you agree?
> 
> The sampled study-specific intercepts should also be used by default for predict() to make study-specific absolute predictions. When passing a baseline distribution, the sampled values from this distribution should be used as .mu, and .mu should not be in newdata. What do you think?

Yes on both of these counts, I think that's the right approach.

ndunnewind commented:
I modified `relative_effects()` and `predict()` as discussed, came up with a more efficient stan implementation, and improved the vignette. I believe this is now almost ready to be merged. @dmphillippo, what do you think?

dmphillippo commented on R/data.R:
```diff
@@ -400,3 +400,22 @@
+
+#' Certolizumab
+#'
+#' Data analysed in \insertCite{TSD3}{multinma}.
```
```suggestion
#' Data frame containing the results of 12 trials comparing 7 treatments for rheumatoid arthritis, analysed in \insertCite{TSD3}{multinma}.
```

dmphillippo commented on R/nma.R:
```diff
@@ -560,6 +561,11 @@ nma <- function(network,
 
+  if (".mu" %in% all.vars(regression)) {
+    if (has_ipd(network)) abort("Cannot fit baseline risk meta-regression model with IPD.")
+    if (QR) abort("Cannot fit baseline risk meta-regression model with QR.")
```
```suggestion
    if (QR) {
      warn("Cannot fit baseline risk meta-regression model with QR decomposition, setting QR = FALSE.")
      QR <- FALSE
    }
```

dmphillippo commented on R/nma.R:
```diff
@@ -708,6 +714,22 @@ nma <- function(network,
+      if (".mu" %in% colnames(idat_agd_arm)) {
+        warn(c(
+          "Detected `.mu` in the `regression` formula and in the data.",
+          i = "`.mu` is a special variable refering to the modelled baseline risk."
```
```suggestion
          "Detected `.mu` in the `regression` formula and in the data.",
          `*` = "`.mu` column in data will be ignored.",
          i = "`.mu` is a special variable referring to the modelled baseline risk."
```

dmphillippo commented on R/nma.R:
```diff
@@ -708,6 +714,22 @@ nma <- function(network,
+
+      idat_agd_arm <- idat_agd_arm %>%
+        dplyr::group_by(.data$.study) %>%
+        dplyr::mutate(.mu = as.integer(.data$.trt != ref_trt & ref_trt %in% .data$.trt)) %>%
```
Do we want instead to put a dummy variable for .mu in every non-reference treatment arm?
```suggestion
        dplyr::mutate(.mu = as.integer(.data$.trt != ref_trt)) %>%
```
And then let interactions with treatment be governed by the setting of `class_interactions`?

dmphillippo commented on R/nma_data.R:
```diff
@@ -1850,3 +1850,54 @@ nfactor <- function(x, ..., numeric = TRUE, resort = FALSE) {
+#'
+#' @param network An `nma_data` object.
+#' @param link The name of the link function to use.
+#' @param treatment The reference treatment. If `NULL` (the default), use the `network`
```
```suggestion
#' @param treatment The reference treatment. If `NULL` (the default), use the network
```

dmphillippo commented on R/nma_data.R:
```diff
@@ -1850,3 +1850,54 @@ nfactor <- function(x, ..., numeric = TRUE, resort = FALSE) {
+  if (!inherits(network, "nma_data")) abort("Not nma_data object.")
+
+  if (!has_agd_arm(network) || has_agd_contrast(network) || has_ipd(network)) {
+    abort("Should only have agd_arm.")
```
```suggestion
    abort("Currently only supported for networks with arm-based aggregate data only.")
```

dmphillippo commented on R/nma_data.R:
```diff
@@ -1850,3 +1850,54 @@ nfactor <- function(x, ..., numeric = TRUE, resort = FALSE) {
+#'   reference treatment.
+#'
+#' @return The baseline risk as a single numeric value.
+#' @export
```
```suggestion
#' @noRd
```

I don't think this function should be exported. The calculation is only crude, and there is no need for the user to call it.

dmphillippo commented on R/nma_data.R:
We still need to:

- [ ] Extend this to IPD sources
- [ ] Extend this to remaining likelihoods (poisson rate, individual binary outcomes, survival)

We might be able to use inbuilt modelling functions (glm, survfit) to do this very generally?

dmphillippo reviewed (commented):
Hi @ndunnewind, thanks so much! Sorry it's taken me so long to get round to looking at this.

Some things we still need to consider:

- [ ] Enable use with IPD
- [ ] Enable all outcomes (poisson rate, individual binary outcomes, survival)
- [x] Work with the `class_interactions` argument, to allow different treatments to have different interactions or to share within classes/all active treatments [Edit: I see this already works - nice!]
- [ ] Add yourself as contributor
- [ ] Add news item

ndunnewind commented on R/nma_data.R:
```diff
@@ -1850,3 +1850,54 @@ nfactor <- function(x, ..., numeric = TRUE, resort = FALSE) {
+    out <- r / n
+
+    # Assume odds of 0.01 if r is 0
+    out[r == 0] <- 0.01 / (1 + 0.01)
```
This is probably not always the best approach. Should we use the +0.5 correction instead, as in the vignette?

dmphillippo commented:
I've just pushed an update for the baseline risk centering code that uses the 0.5 continuity correction and works for all outcomes and IPD (survival the centering is just returned as 0, with a note).

For the failing DIC tests, I believe that our results are consistent across platforms and that the TSD output is slightly off - some were falling just the wrong side of the tolerance threshold to fail the test.

I just need to add your code to the survival .stan files and run a few more tests and then I think we should be nearly good to go? Make sure you're happy with the tweaks I've made too! (e.g. I simplified the addition of a dummy .mu column to the data, since the make_nma_design_matrix() function handles the necessary contrasts, and now this works properly when not every study has the reference treatment.)

dmphillippo commented:
Pending the tests passing, I think this is now ready to go! Take a look and see if you are happy. I've added a news item:

* Feature: Regression on baseline risk is now supported, with a new `.mu` special variable for `regression` formulas which can interact with treatment, e.g. `regression = ~.mu:.trt`. Following TSD3, the regression is performed on the estimated intercept parameter not the observed baseline risk, to correctly capture uncertainty and account for correlations between baseline risk and treatment effect. A new Certolizumab vignette demonstrates the new features. Thanks to @ndunnewind (PR #36).

ndunnewind commented on R/nma_data.R:
```diff
@@ -1850,3 +1850,72 @@ nfactor <- function(x, ..., numeric = TRUE, resort = FALSE) {
+      if (has_ipd(network)) dplyr::group_by(network$ipd, ".study", ".trt") %>%
+        dplyr::summarise(.r = sum(.data$.r[,2]), .n = dplyr::n()),
+      if (has_agd_arm(network)) dplyr::select(network$agd_arm, ".study", ".trt", ".r") %>%
+        dplyr::mutate(.n = rowSums(.data$.r), .r = .data$.r[,2])
```
The data are not in inclusive format here, so I think `.r` should be the sum of all columns expect the first one. Also, categories may be missing, so we should probably use `na.rm = TRUE`.

```suggestion
      if (has_ipd(network)) dplyr::group_by(network$ipd, ".study", ".trt") %>%
        dplyr::summarise(.r = rowSums(.data$.r[, -1], na.rm = TRUE), .n = dplyr::n()),
      if (has_agd_arm(network)) dplyr::select(network$agd_arm, ".study", ".trt", ".r") %>%
        dplyr::mutate(.n = rowSums(.data$.r, na.rm = TRUE), .r = rowSums(.data$.r[, -1], na.rm = TRUE))
```

dmphillippo replied:
Ah yes, you're right! Baseline risk in this case is the event probability for the first outcome, but everyone achieving the higher outcomes achieves this too, so we want to sum the outcomes. Agreed on the na.rm use too.

ndunnewind commented on R/nma.R:
```diff
@@ -795,6 +817,18 @@ nma <- function(network,
+        ))
+      }
+
+      idat_agd_arm$.mu <- 1L
```
For studies without a reference treatment arm, the study intercept is not the baseline risk. So should we not set .mu to zero so that the baseline risk meta-regression does not affect such studies?

```suggestion
      ref_trt <- levels(idat_agd_arm$.trt)[1L]

      idat_agd_arm <- idat_agd_arm %>%
        dplyr::group_by(.data$.study) %>%
        dplyr::mutate(.mu = as.integer(ref_trt %in% .data$.trt)) %>%
        dplyr::ungroup()
```

dmphillippo replied:
No, I don't think so - when `class_interactions` is not `"common"` the interactions with baseline risk will not cancel out, because the interactions for each non-reference treatment are different, so we need to include the BRMR terms on all arms. We use a model parameterisation where the intercept in every study is the network reference treatment ([described in this paper](https://doi.org/10.1111/rssa.12579)), so this is the right baseline risk parameter.

ndunnewind replied:
@dmphillippo This still needs some attention before merging. I think models are incorrectly specified now for networks containing head to head comparisons.

dmphillippo replied:
@ndunnewind I think these are correct as they are? 

The intercept `mu` always corresponds to the network reference treatment in the parameterisation that we use. If the reference treatment is A, and there is a two-arm study comparing B and C, then the NMA model that we fit is

$$
\textrm{Arm B: } \mu^{(A)} + d_B
$$

$$
\textrm{Arm C: } \mu^{(A)} + d_C
$$

The meta-regression on baseline risk (with `class_interactions = "independent"`) is then

$$
\textrm{Arm B: } \mu^{(A)} + \beta_B (\mu^{(A)} - \bar{\mu}) + d_B
$$

$$
\textrm{Arm C: } \mu^{(A)} + \beta_C (\mu^{(A)} - \bar{\mu}) + d_C
$$

This may not always be estimable, but it is correct! The interactions of baseline risk for treatments B and C do not necessarily cancel out. This is the "Model 1. Independent, treatment-specific interactions" in Box 1 (page 31) of TSD 3.

With `class_interactions = "common"` and an appropriate class factor set so that $\beta_B = \beta_C$, we get the more usual simplified model in TSD 3 which does cancel out the interactions on the active-active comparisons ("Model 3. The same interaction for all treatments" in TSD 3 Box 1).

Note also that TSD 3 (p43) says

> [Model 3, the common interactions model] assumes the same interaction effect for all treatments compared to treatment 1, the regression terms will cancel for all other comparisons, so no baseline risk adjustment is performed for trials which do not include treatment 1. If fitting one of the other models in Box 1, care should be taken to ensure that the risk being adjusted for refers to the estimated risk for the reference treatment (treatment 1) which may not have been compared in every trial.

which we are indeed following (`mu` always corresponds to the network reference treatment, even in trials without this treatment).

These are the models that I hope we have implemented anyway, unless I'm missing something! Looking at the code I think this is right 🙂

ndunnewind replied:
Thanks for the detailed explanation! I agree this makes sense if the study intercept is always the network reference treatment.

ndunnewind commented on R/nma.R:
```diff
@@ -2888,6 +2934,9 @@ make_nma_formula <- function(regression,
       nma_formula <- regression
     }
 
+    # Remove any main effect of .mu if baseline risk regression used
+    if (".mu" %in% all.vars(regression)) nma_formula <- update.formula(nma_formula, ~. - .mu)
```
Maybe abort here to prevent confusion?

```suggestion
    if (".mu" %in% attr(terms(regression), "term.labels")) {
      abort("`.mu` is not allowed as main effect in the `regression`.")
    }
```

dmphillippo replied:
I silently removed any main effects, because this is how main effects of `.trt` and `.trtclass` are handled too. The "main effects" of `.mu` are still in the model in a way - they are the intercepts. The proposed change gives an error for `~.mu*.trt`, which would be nice to silently correct (the same way that the main effects of `.trt` are handled in this case). 

If we can distinguish between `~.mu*.trt` and `~.mu + ...` then we could add a warning that the main effects are being removed for the latter? This would probably need to go in `nma()` rather than `make_nma_formula()`, to avoid this error firing when `make_nma_formula()` is called in predict etc.

ndunnewind replied:
Then it probably makes sense to leave it like this for now, and maybe change this for `.trt` and `.trtclass` as well in separate PR.

dmphillippo commented:
@ndunnewind I think this is ready to merge, if you are happy?

dmphillippo commented:
Apologies, I've only just realised that my comments to your previous replies were likely not visible - I had to go to the files tab to actually "submit my review" to send these. Sorry!

Thanks again!

ndunnewind commented:
No worries! I am happy with all changes now.

dmphillippo commented:
Brilliant, thank you so much! Merging now
